Add SimPhotonCamera

gerth2 opened the pull request:
~RFC style draft architecture. I started in Java only for discussion purposes, and will port to C++ once we're happy with the results.~ Updated - we should be good to go now. See final results below.

Exposes a new `SimPhotonCamera` object to let a user inject simulation results into NT, emulating a real photonvision software instance running on a coprocessor.

Inherit from the base `PhotonCamera` to get a similar set of functionality. 

For sim, in each update loop, the user is expected to:

1) Call `reportDetectedTarget()` once for each visible target, with information about where the target should be in the camera's field of view
2) Call `update()` to post all target data to networkTables.

`update()` only posts if the camera is not in driver mode.

Unmodeled:
1) How and where the target appears in the image - user must provide all this information.
2) How LED state impacts target visibility
3) How pipeline impacts visibility
4) Impact of Camera FOV on measurements and visibility.
5) Latency (hardcoded to zero currently)

An additional simulation layer we could provide (or at least show an example for) would take in configuration for known target Pose2d's and camera capabilities (FOV, range, LED's, etc.). Then, at runtime, given the robot's Pose2d, it could determine which targets would be in view and call the appropriate API's in `SimPhotonCamera`. IMO I think you'd want to write this to target the "90%" usecase and get teams started, and punt on the 10% of wacko-usecase teams.

gerth2 committed:
Added new API's to control camera image save-to-file functionality

gerth2 committed:
Merge branch 'master' into master

gerth2 committed:
Fixing formatting

gerth2 committed:
Second pass at format cleanup

gerth2 committed:
Revised API to simpler "takeSnapshot" formats.

This largely grew out of discord discussions and the PR.

gerth2 committed:
Formatting fixups.

https://www.youtube.com/watch?v=CduA0TULnow

mcm001 committed:
Merge branch 'master' into master

gerth2 committed:
WIP starting a simulation photoncamera

gerth2 committed:
Merge branch 'master' of https://github.com/gerth2/photonlib into master

gerth2 committed:
Import Cleanup

gerth2 committed:
Merge remote-tracking branch 'upstream/master' into master

# Conflicts:
#	src/main/java/org/photonvision/PhotonCamera.java

mcm001 commented on src/main/java/org/photonvision/SimPhotonCamera.java:
```diff
@@ -0,0 +1,91 @@
+    
+    private Packet packet = new Packet(1);
+
+    LinkedList<PhotonTrackedTarget> trackedTargetList = new LinkedList<PhotonTrackedTarget>();
```
When is this list ever cleared?

gerth2 replied:
should be at the end of the `update()` method

That being said - per discord conversation - I think we can circumvent this whole thing by going to vararg list of targets which get passed straight to `update()` and not need the list anymore?

gerth2 committed:
Added driverMode catch for not publishing target info

gerth2 committed:
Test unused import cleanups

gerth2 committed:
Refactored per discord conversations

gerth2 committed:
Applied Formatter

gerth2 commented:
More details after refactor:

`SimPhotonCamera` now has a single `submitProcessedFrame()` API where users can submit a set of `PhotonTrackedTarget`'s to network tables

All the real calculations are now happening in `SimVisionSystem`. In addition to specifying a whole lot of properties about their camera and pipline settings, `addSimVisionTarget` gets called once per target on the field that should be identifiable. `SimVisionTarget`'s are just data-organization objects to hold properties about the size and location of targets on the field.

At runtime, the `processFrame()` api gets called with the robot pose as the only argument, and all the other math becomes internal. This includes a four-part check of visibility:

1. Within vertical FOV
2. Within horizontal FOV
3. Large enough to pass through contour filtering
4. Close enough to be illuminated by LED's

gerth2 commented:
TODO Stilll:

0. ~~Confirm folks like it~~
1. Refactor post-monorepoing
2. ~~Add unit tests~~
3. ~~rinse wash repeat but with C++~~

mcm001 commented on src/main/java/org/photonvision/SimVisionSystem.java:
```diff
@@ -0,0 +1,139 @@
+     * @param cameraRes_px Pair of integers representing the Width and Height (respectively) of your camera's image sensor in pixels
+     * @param minTargetArea Minimum area that that the target should be before it's recognized as a target by the camera. Match this with your contour filtering settings in the PhotonVision GUI.
+     */
+    public SimVisionSystem(String camName, double camDiagFOV_deg, double camPitch_deg, Transform2d robotToCamera, double cameraHeightOffGround_m, double maxLEDRange_m, Pair<Integer, Integer> cameraRes_px, double minTargetArea){
```
Change the pair of integers to a Size?

mcm001 replied:
and _m to Meters?

mcm001 replied:
idk thats what wpilib does

gerth2 replied:
10-4, I'll confirm whatever they have and follow suit. the "Pair" was an arbitrary invention and choice on my end.

gerth2 replied:
FWIW I'd attempted to avoid using openCV types as part of public API's, but that's really the only driving reason I had not to do `Pair`. "Match wpilib" is still my thought for best strategy.

mcm001 commented on src/main/java/org/photonvision/SimVisionSystem.java:
```diff
@@ -0,0 +1,139 @@
+
+        Pose2d cameraPos = robotPose.transformBy(robotToCamera);
+
+        LinkedList<PhotonTrackedTarget> visibleTgtList = new LinkedList<PhotonTrackedTarget>();
```
why LinkedList? Just curious.

gerth2 replied:
Picked for constant-time add operation. However, thinking through it again, an `ArrayList` init'ed to the same size as the total number of possible targets would perform better. Assuming (I think) execution time is much more important than memory.

mcm001 commented on src/main/java/org/photonvision/SimVisionSystem.java:
```diff
@@ -0,0 +1,139 @@
+        boolean inHorizAngle = Math.abs(yaw) < (this.camHorizFOV_deg/2);
+        boolean inVertAngle = Math.abs(pitch) <  (this.camVertFOV_deg/2);
+        boolean targetBigEnough = area > this.minTargetArea;
+        return (inRange & inHorizAngle & inVertAngle & targetBigEnough);
```
Should these be `&&`?

gerth2 replied:
Yes https://www.tutorialspoint.com/Java-Boolean-operators

mcm001 commented on src/main/java/org/photonvision/SimVisionTarget.java:
```diff
@@ -0,0 +1,44 @@
+     * @param targetHeight_m Height of the target above the field plane, in meters.
+     * @param targetDims_m Pair of floating point values describing the width and height of the outter bounding box of the target. Both are in meters.
+     */
+    public SimVisionTarget(Pose2d targetPos, double targetHeight_m, Pair<Double, Double> targetDims_m){
```
We could also use OpenCV's `Size` here.

gerth2 replied:
I believe this will be the same as above.

gerth2 committed:
Review comment updates

gerth2 committed:
Test submitting a simple unit test to CI to see what happens

gerth2 committed:
more test test test

gerth2 committed:
Test all the java things!

gerth2 committed:
WIP C++ skeleton

gerth2 committed:
Even more C++ WIP

gerth2 committed:
Merge remote-tracking branch 'upstream/master' into master

gerth2 committed:
Even more c++ wip.

gerth2 committed:
At long last, it builds!

gerth2 committed:
c++ argument type fixups and first simple test added.

gerth2 committed:
Merge remote-tracking branch 'upstream/master' into master

# Conflicts:
#	build.gradle

gerth2 committed:
wip with simpiler unit tests

gerth2 committed:
Reordered PhotonCamera.h order, works now.

KJSDFHFKJSDFHSDKJFHSDKJFHKSDJHFKLSDJFHKDSJHFKjh

gerth2 committed:
nvm yes we do need the build.gradle updates, otherwise java unit sims dont run.

gerth2 committed:
Cleanup and first actual meaningful unit sim. Still WIP though.

gerth2 committed:
Per advice from Peter, we're up and running with a workaround, though still failing the unit sim. Punting for tonight.

gerth2 committed:
Hot diggity dog I think it works finally

gerth2 committed:
Do not need this log file I think.

gerth2 committed:
Make format _actually_ happy.

gerth2 committed:
Another test added

gerth2 committed:
Completed porting java unit tests to c++

gerth2 committed:
Fix CI - apparently all non-windows compilers can't figure out how to turn "11" into an unsigned long int themselves?

prateekma commented on src/main/native/cpp/photonlib/PhotonCamera.cpp:
```diff
@@ -45,10 +45,16 @@ PhotonPipelineResult PhotonCamera::GetLatestResult() const {
-  packet >> result;
 
-  // Return result
+  std::vector<char> bytes;
+
+  for (char ch : rawBytesEntry.GetValue()->GetRaw()) {
+    bytes.push_back(ch);
+  }
```
Can be replaced by:
```
std::string value = rawBytesEntry.GetValue()->GetRaw();
std::vector<char> bytes{value.begin(), value.end()};
```

prateekma commented on src/main/native/cpp/photonlib/SimPhotonCamera.cpp:
```diff
@@ -0,0 +1,42 @@
+SimPhotonCamera::SimPhotonCamera(const std::string& cameraName)
+    : PhotonCamera(cameraName) {}
+
+void SimPhotonCamera::submitProcessedFrame(
```
Rename to `SubmitProcessedFrame`.

prateekma commented on src/main/native/cpp/photonlib/SimVisionSystem.cpp:
```diff
@@ -0,0 +1,122 @@
+ */
+
+#include "photonlib/SimVisionSystem.h"
+
```
Add includes for `<units/length.h>` and `<units/angle.h>`.

gerth2 replied:
Do you want them in both the .cpp and .h file?

prateekma replied:
Ah I missed that it was in the header. Just the .h is good enough.

prateekma commented on src/main/native/cpp/photonlib/SimVisionSystem.cpp:
```diff
@@ -0,0 +1,122 @@
+namespace photonlib {
+
+SimVisionSystem::SimVisionSystem(
+    const std::string& name, units::angle::degree_t camDiagFOVDegrees,
+    units::angle::degree_t camPitchDegrees, frc::Transform2d robotToCamera,
+    units::length::meter_t cameraHeightOffGroundMeters,
+    units::length::meter_t maxLEDRangeMeters, int cameraResWidth,
```
You can just do `units::meter_t` and `units::degree_t`.

prateekma replied:
You don't need to append the parameter names with the units -- that is only necessary in Java.

prateekma commented on src/main/native/cpp/photonlib/SimVisionSystem.cpp:
```diff
@@ -0,0 +1,122 @@
+  frc::Pose2d cameraPos = robotPoseMeters.TransformBy(robotToCamera);
+  std::vector<PhotonTrackedTarget> visibleTgtList = {};
+
+  std::vector<SimVisionTarget>::iterator tgt;
+  for (tgt = tgtList.begin(); tgt != tgtList.end(); ++tgt) {
```
`for (auto&& tgt : tgtList)`

prateekma commented on src/main/native/cpp/photonlib/SimVisionSystem.cpp:
```diff
@@ -0,0 +1,122 @@
+      procDelay, wpi::MutableArrayRef<PhotonTrackedTarget>(visibleTgtList));
+}
+
+double SimVisionSystem::getM2PerPx(units::length::meter_t dist) {
```
Method name should start with uppercase letter.

prateekma commented on src/main/native/cpp/photonlib/SimVisionSystem.cpp:
```diff
@@ -0,0 +1,122 @@
+  return widthMPerPx * heightMPerPx;
+}
+
+bool SimVisionSystem::camCanSeeTarget(units::length::meter_t distMeters,
```
Same as above.

gerth2 committed:
Fixed up code review comments:

-- Removed redundant units in variable names
-- Proper capitalization of method names
-- Proper header file includes
-- Modernized some for loops to better idioms

gerth2 committed:
Switched units initializers to use _m or _deg suffix methodologies.

Swiched as many of the ASSERT statements to be EXCEPT as possible.

mcm001 commented on src/main/java/org/photonvision/PhotonUtils.java:
```diff
@@ -60,4 +60,17 @@ public static double calculateDistanceToTargetMeters(double cameraHeightMeters,
+   * @param angle
+   * @return Wrapped-to-range angle
+   */
+  static double wrapAngleDeg(double angle){
+    angle %=360;
+    angle = angle > 180.0 ? angle-360 : angle;
+    angle = angle <= -180.0 ? angle+360 : angle;
+    return angle;
+  }
```
Replace with MathUtil.inputModulus?

gerth2 committed:
Removing unused function

gerth2 committed:
Revised api to use cameraToRobot transform (rather than cameraToRobot)

mcm001 committed:
Merge branch 'master' into master

gerth2 committed:
Fixed formatting
